sklearn/cluster/optics_.py

```diff
@@ -527,3 +555,379 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+        indices inclusive. The clusters are ordered in a way that larger
+        clusters encompassing smaller clusters come after those smaller
+        clusters.
+    """
+
+    # all indices are inclusive (specially at the end)
+    n_samples = len(reachability_plot)
+    # add an inf to the end of reachability plot
+    # but process the data only until the last actual point
+    # this is fine since the last point is considered upward anyway
+    reachability_plot = np.hstack((reachability_plot, np.inf))
+
+    if min_cluster_size <= 1:
+        min_cluster_size = max(2, min_cluster_size * n_samples)
+
+    xi_complement = 1 - xi
+    sdas = []  # steep down areas, introduced in section 4.3.2 of the paper
+    clusters = []
+    index = 0
+    mib = 0.  # maximum in between
+
+    ratio = reachability_plot[:-1] / reachability_plot[1:]
+    steep_upward = ratio <= xi_complement
+    steep_downward = ratio >= 1 / xi_complement
+    downward = ratio > 1
+    upward = ratio < 1
+
+    # the following loop is is almost exactly as Figure 19 of the paper.
+    # it jumps over the areas which are not either steep down or up areas
+    for steep_index in iter(np.flatnonzero(steep_upward | steep_downward)):
+        # just continue if steep_index has been a part of a discovered xward
+        # area.
+        if steep_index < index:
+            continue
+
+        mib = max(mib, np.max(reachability_plot[index:steep_index + 1]))
+        index = steep_index
+
+        # check if a steep downward area starts
+        if steep_downward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
+            D_start = index
+            index, end = _extend_downward_region(steep_downward, upward,
+                                                 D_start, min_samples)
+            D = _Area(start=D_start, end=end,
+                      maximum=reachability_plot[D_start], mib=0.)
+            sdas.append(D)
+            mib = reachability_plot[end + 1]
+
+        elif steep_upward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
+            U_start = index
+            index, end = _extend_upward_region(steep_upward, downward, U_start,
+                                               min_samples)
+            U = _Area(start=U_start, end=end, maximum=reachability_plot[end],
+                      mib=-1)
+            mib = reachability_plot[end + 1]
+
+            U_clusters = []
+            for D in sdas:
+                c_start = D.start
+                c_end = min(U.end, n_samples - 1)
+
+                # line (**), sc2*
+                if reachability_plot[c_end + 1] * xi_complement < D.mib:
+                    continue
+
+                # Definition 11: criterion 4
+                if D.maximum * xi_complement >= reachability_plot[c_end + 1]:
+                    # Find the first index from the left side which is almost
+                    # at the same level as the end of the detected cluster.
```

Comment:
that's the language of the paper, but I find it vague (was the meaning of "almost at the same level" defined in the paper?). Here we're finding the first index from the left side that is below the end of the cluster in the reachability plot, aren't we?

Reply:
more like, we're finding the last index from the left side which is higher than the reachability of the point after the end of the cluster. The paper motivates it by saying what I have as the comment. I'm not sure if writing what the while loop does, in words, would help here.